### DISCUSSION TITLE
Inline Julia code to define fig-cap using Jupyter kernel and Quarto 1.4 pre-release?

### DISCUSSION THREAD

**Discussion by dcnorris** [quarto-dev/quarto-cli#6369] [Q&A]
### Description

I see that the 1.4 pre-release [supports](https://quarto.org/docs/prerelease/1.4/inline.html) "executing inline expressions when using Jupyter kernels."  Indeed, I can get Julia code to execute inside the main text:

```
(Mean ESS across all parameters: `{julia} @sprintf "%3.f" mean_ess`.)
```
[Please don't @ me about computing a mean ESS ;).]

But I cannot make this work to obtain a computed `fig-cap` setting.  I've seemingly exhausted the possible syntactical permutations, and suspect this is not yet supported for Julia; would anyone confirm or deny this?

**Comment by cscheid** [comment]
> But I cannot make this work to obtain a computed `fig-cap` setting. 

Can you share a full, minimal example?

**Comment by dcnorris** [comment]
(I'm using quarto v1.4.267.)

````
---
title: "Trouble computing fig-cap with Julia"
author: David C. Norris
format:
  html:
    code-fold: true
    self-contained: true
jupyter: julia-1.9
---

```{julia}
n = 10
using Plots
```

### This first plot caption comes out saying:

> {julia} "Plot of numbers 1 thru $n."

```{julia}
#| label: too-bad
#| fig-cap: '`{julia} "Plot of numbers 1 thru $n."`'

plot(1:n)
```

### The next 2 plots get captions that read:

> [object Object]

```{julia}
#| label: too-bad-2
#| fig-cap: !expr '{julia} "Plot of numbers 1 thru $n."'

plot(1:n)
```

```{julia}
#| label: too-bad-3
#| fig-cap: !expr '`{julia} "Plot of numbers 1 thru $n."`'

plot(1:n)
```
````

**Comment by cscheid** [comment]
Thanks. I just wanted to make sure.

> #| fig-cap: '`{julia} "Plot of numbers 1 thru $n."`'

This is not a feature that exists. (Did we imply this feature exists in our docs? If so, we should fix that.)

**Reply by mcanouil** [reply]
To date, the only piece of documentation regarding this two weeks old feature is <https://quarto.org/docs/prerelease/1.4/inline.html>.
And some parts are missing such as how to escape the inline commands for educational purposes.

**Comment by dcnorris** [comment]
Your docs didn't claim this feature exists, but they also didn't say it doesn't.  Since this sort of thing *is* possible with R, perhaps you would do well to clarify it's not yet implemented for Julia.

**Reply by cscheid** [reply]
> Since this sort of thing is possible with R

I hear you. Unfortunately, `!expr` is a knitr feature that we can't disable, and we explicitly don't document it in quarto (and we try to nudge users away from).

What are you trying to do more generally? There might be a way around it.

**Reply by mcanouil** [reply]
@cscheid can confirm/develop, but it's not certain that code evaluation within YAML will propagate as it currently does through `knitr` which is in fact more of a side effect rather than a Quarto feature.

**Reply by baptiste** [reply]
This feature is common enough that I think it'd be nice to have a language-agnostic way to obtain it. I wonder if one could use an analogy with ggplot's build process: some steps need to hold on until a later step is performed in the toolchain, to get a definite value. For instance, scales need to be trained on the data before the scale limits are known, and this requires more than one pass. The implementation considers _promises_ such as [waiver()](https://ggplot2.tidyverse.org/reference/waiver.html), which act as placeholders to be filled when actual values are known.

In the context of chunk options referring to computed values, this would mean:

- introducing a syntax for such "promises", e.g. `#| fig-cap: '~{"Plot of numbers 1 thru $n."}'`
- these would be flagged as needing interpolation after the chunk has been executed by the engine
- the convention could be to only evaluate within the chunk environment, to make things explicit 
- interpolation could be implemented in a single language, possibly lua or ojs, and occur just at the start of pandoc's AST processing: look for the flagged items, and use a dictionary from the objects defined in the chunk environment to inject the variables. As I understand, this is (maybe?) more or less what's considered for [inline code](https://github.com/quarto-dev/quarto-cli/discussions/5638#discussioncomment-5982952).

**Reply by cscheid** [reply]
> This feature is common enough that I think it'd be nice to have a language-agnostic way to obtain it.

I 100% agree. We have a design, but only so many hours in the day...

**Reply by baptiste** [reply]
That's nice to hear – the message above gave me the impression that this was not in the future plans (and might even be retroactively removed someday, in the case of knitr).

**Reply by mcanouil** [reply]
@baptiste It was mostly about the `!expr` specific `knitr` inline evaluation. (Or the `` '`r ...`' ``)

**Reply by dcnorris** [reply]
Referencing $N$'s in figure captions is a basic requirement for reproducible research.  Is there a workaround one might use in the meantime?

**Reply by baptiste** [reply]
You could use custom LaTeX macros to get around this,

````
---
title: "Untitled"
format: html
---

```{r}
#| echo: false
#| results: asis
N <- 1234
cat(paste0("$\\newcommand{\\mycaptionN}{",N,"}$"))
```


```{r}
#| label: too-bad
#| fig-cap: The value is $\mycaptionN$
plot(1:3)
```

````

Obviously a hack, but I think the approach could be extended via Lua filters (or something built-in), basically having placeholders that are replaced once the variables are known. In this case pandoc is doing the macro substitution, as a side-effect of how it handles maths.

**Reply by mcanouil** [reply]
@baptiste could you edit your example to only use one code cell options syntax, preferably the YAML syntax recommended in Quarto?

**Reply by cderv** [reply]
> Your docs didn't claim this feature exists, but they also didn't say it doesn't. Since this sort of thing is possible with R, perhaps you would do well to clarify it's not yet implemented for Julia.

@dcnorris Regarding our doc, we currently only mention in the R computation doc the `!expr` syntax as a way to use R expression in YAML 
https://quarto.org/docs/computations/r.html#chunk-options

We could try to clarify that this is really not working for other language, but by the fact it is in the `engine: knitr` doc  only we though it was clear it is not working for others like `engine: jupyter`. Do you have a suggestion ?

**Reply by dcnorris** [reply]
My attempts to use `!expr` were just hacking away, hoping something would work.  Pending the actual release of 1.4, you might just explicitly state — at the point where you mention this new feature https://quarto.org/docs/prerelease/1.4/inline.html — that it cannot be used in YAML strings.
